### PR TITLE
Fix quit on ctrlc, race panic, atomic load align in session IO

### LINKED ISSUE #11092: CTRL-C causing sessions to freeze on 9.0
### Description

After installing teleport 9.x, when I attempt to send a ctrl-c to the remote terminal the session completely hangs and I have to use the ssh escape sequence ~. to quit and reconnect

**What you expected to happen**:
ctrl-c should be passed through without breaking.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
Simply send ctrl-c to the console

### Server Details
- Teleport version (run `teleport version`):
Teleport v9.0.0 git:v9.0.0-0-g1fa8857aa go1.17.7
- Server OS (e.g. from `/etc/os-release`):
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware):
local, inside proxmox
- Additional details:

### Client Details
- Tsh version (`tsh version`):
Teleport v9.0.0 git:v9.0.0-0-g1fa8857aa go1.17.7
- Computer OS (e.g. Linux, macOS, Windows):
Ubuntu 20.04
- Browser version (for UI-related issues):
N/A
- Installed via (e.g. apt, yum, brew, website download):
APT
- Additional details:

### Debug Logs
Please include or attach debug logs, when appropriate. Obfuscate sensitive information!
- Start Teleport with --debug flag (`teleport --debug`)
- Run tsh with --debug flag (`tsh --debug`)

0 log output after connection establishes until such time as I kill the session with the escape sequence.

### LINKED ISSUE #11106: Teleport panics with unaligned 64-bit atomic operation on 32bit ARM
### Description

**What happened**:
I tried adding a 32bit ARM node (raspberry pi running a customized OpenWRT) to a teleport cluster,
following the instructions for [Adding a Node located behind NAT](https://goteleport.com/docs/setup/admin/adding-nodes/#adding-a-node-located-behind-nat)
The teleport binary crashes reliably as soon as I try to connect to it via the webinterface or tsh.

The session in the webinterface ends with:
```
disconnected
wait: remote command exited without exit status or exit signal
```

On the ARM node teleport crashes with:
```
DEBU [NODE]      Successfully authenticated fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:8342ZwDCXygxbXC3am1BFZe2JGPW3rH5FxeLBmxX+eg local:192.168.42.27:44350 remote:192.168.42.54:39892 user:root srv/authhandlers.go:293
DEBU [NODE]      Checking permissions for (testuser,root) to login to node with RBAC checks. srv/authhandlers.go:422
DEBU [RBAC]      Access to node "acd7cdeb-2046-4508-a768-50baf4e98878" granted, allow rule in role "access" matched. services/role.go:1853
DEBU [SSH:NODE]  Incoming connection 192.168.42.54:39892 -> 192.168.42.27:44350 version: SSH-2.0-Go, certtype: "user" sshutils/server.go:466
DEBU [SSH:NODE]  Received out-of-band request: &{Type:x-teleport-version WantReply:true Payload:[] ch:<nil> mux:0x71b9d80}. sshutils/server.go:521
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [KEEPALIVE] Starting keep-alive loop with with interval 5m0s and max count 3. srv/keepalive.go:65
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Handling request env, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU             Will create new session for SSH connection 192.168.42.54:39892. srv/ctx.go:512
DEBU [NODE]      Handling request auth-agent-req@openssh.com, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Starting agent server for Teleport user testuser and socket /tmp/teleport-2905853319/teleport-12593.socket. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:889
DEBU             Will create new session for SSH connection 192.168.42.54:39892. srv/ctx.go:512
DEBU [NODE]      Handling request pty-req, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [NODE]      Requested terminal "xterm-256color" of size {80 25} id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser srv/termhandlers.go:73
DEBU             Set permissions on /dev/pts/1 to 0:0 with mode -rw--w----. srv/term.go:407
DEBU [SESSION:N] Unable to update window size, no session found in context. srv/sess.go:445
DEBU             Will create new session for SSH connection 192.168.42.54:39892. srv/ctx.go:512
DEBU [NODE]      Handling request shell, want reply true. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser regular/sshserver.go:1398
DEBU [SESSION:N] Creating tracker srv/sess.go:1709
INFO [NODE]      Creating (interactive) session 1b91eb19-3efe-40f0-92c1-c556ce6ffe09. id:1 local:192.168.42.27:44350 login:root remote:192.168.42.54:39892 teleportUser:testuser srv/sess.go:223
DEBU [SESSION:N] Using async streamer for session 1b91eb19-3efe-40f0-92c1-c556ce6ffe09. srv/sess.go:1255
DEBU [SESSION:N] Tracking participant: testuser srv/sess.go:1751
INFO [AUDIT]     session.start addr.remote:192.168.42.54:39892 cluster_name:PL911 code:T2000I ei:0 event:session.start login:root namespace:default server_addr:192.168.42.27:44350 server_hostname:testberry server_id:acd7cdeb-2046-4508-a768-50baf4e98878 session_recording:node sid:1b91eb19-3efe-40f0-92c1-c556ce6ffe09 size:80:25 time:2022-03-14T12:03:47.144Z uid:315481ea-e10a-4f13-814a-7878e3e5db7e user:testuser events/emitter.go:325
INFO [SESSION:N] New party ServerContext(192.168.42.54:39892->192.168.42.27:44350, user=root, id=1) party(id=d995772a-1da4-4c6a-b935-d5b68efaea5b) joined session: 1b91eb19-3efe-40f0-92c1-c556ce6ffe09 srv/sess.go:1516
DEBU             Will join session 1b91eb19-3efe-40f0-92c1-c556ce6ffe09 for SSH connection 192.168.42.54:39892. srv/ctx.go:514
DEBU [SESSION:N] Launching session 1b91eb19-3efe-40f0-92c1-c556ce6ffe09. srv/sess.go:823
panic: unaligned 64-bit atomic operation

goroutine 241 [running]:
runtime/internal/atomic.panicUnaligned()
	/opt/go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x78cecd4, 0x0)
	/opt/go/src/runtime/internal/atomic/atomic_arm.s:256 +0x14
github.com/gravitational/teleport/lib/srv.(*TermManager).writeToClients(0x78cecc0, {0x0, 0x0, 0x0})
	/go/src/github.com/gravitational/teleport/lib/srv/termmanager.go:83 +0x140
github.com/gravitational/teleport/lib/srv.(*TermManager).writeUnconditional(...)
	/go/src/github.com/gravitational/teleport/lib/srv/termmanager.go:169
github.com/gravitational/teleport/lib/srv.(*TermManager).On(0x78cecc0)
	/go/src/github.com/gravitational/teleport/lib/srv/termmanager.go:192 +0x94
github.com/gravitational/teleport/lib/srv.(*session).launch(0x78f0270, 0x756d080)
	/go/src/github.com/gravitational/teleport/lib/srv/sess.go:827 +0x1b8
github.com/gravitational/teleport/lib/srv.(*session).addParty.func2(0x78f0270)
	/go/src/github.com/gravitational/teleport/lib/srv/sess.go:1548 +0x24
created by github.com/gravitational/teleport/lib/srv.(*session).addParty
	/go/src/github.com/gravitational/teleport/lib/srv/sess.go:1547 +0x680
```

**What you expected to happen**:
A successful connection to the node with shell access.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Download the `Linux ARMv7 (32-bit)` release from https://goteleport.com/teleport/download/
2. Unpack it and copy it onto a 32bit ARM system e.g. a RaspberryPi running a 32bit Raspbian
3. Generate an access token for your cluster e.g. `tctl nodes add --roles=node`
4. Run teleport on the 32bit ARM node e.g. `./teleport start  --roles=node --token=<token> --ca-pin=<capin> --auth-server=<authserver>:3080 --insecure -d`
5. Try to connect to the node you've just added. 

### Node Details
- Node OS: Customized OpenWRT
- Teleport version: `v9.0.0 git:v9.0.0-0-g1fa8857aa go1.17.7`
- Hardware: Raspberry Pi

### Server Details
- Server OS: Ubuntu 20.04
- Teleport version: `v9.0.0 git:v9.0.0-0-g1fa8857aa go1.17.7`
- Where are you running Teleport? Dedicated Hardware (Proxy and Authserver running on my desktop machine for testing)

### Client Details
- Tsh version: Teleport v9.0.0 git:v9.0.0-0-g1fa8857aa go1.17.7
- Computer OS:  Ubuntu 20.04
- Installed via website download (extracted from tarball)

### Additional Information
As far as I can tell this bug is causes by an incorrect usage of `atomic.AddUint64` in [lib/srv/termmanager.go#L83](https://github.com/gravitational/teleport/blob/v9.0.0/lib/srv/termmanager.go#L83).

The [documentation]() for `sync/atomic` states:
```
On ARM, 386, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. 
The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned.
```

Looking at [TermManager](https://github.com/gravitational/teleport/blob/v9.0.0/lib/srv/termmanager.go#L33) struct,
`countWritten` is not within the first word of the struct.
So to fix this issue it should be sufficient to move `countWritten` and `countRead` to the beginning of the struct,
to keep them 64bit aligned.

### PR DESCRIPTION
**Pull request by xacrimon**
`lib/srv/termmanager.go` is responsible for dealing with most session IO and handles behaviours like multiplexing. This PR fixes the following bugs:
- Unaligned integers being used as atomics #11106 
- CTRL-C causing the reader goroutine to quit early https://github.com/gravitational/teleport/issues/11092
- A race condition that could case a panic since we didn't synchronise and check the correct flags prior to sending over a potentially closed channel.

This PR does remove some usage of atomics in termmanager and instead locks additional fields under the mutex. I didn't observe any noticable performance drop here and this would only become a problem if you had hundreds of users typing in a session at the same time. This makes the code somewhat easier to reason about but we'll have to look into that again if it becomes a problem at some point in the future.

### CODE REVIEW

**Comment by rosstimothy** [comment]
It doesn't look like `lib/srv/termmanager.go` has much test coverage at all. Could we at least add some for the issues this is fixing to prevent any regressions here?

File: lib/srv/termmanager_test.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"time"
+	"unsafe"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAtomicAlign(t *testing.T) {
+	m := NewTermManager()
+
+	verifyAlign := func(field *uint64) {
```
Comment (zmb3) [inline lib/srv/termmanager_test.go:31]: nice idea!

**Review by zmb3 — APPROVED** [review]

File: lib/srv/termmanager_test.go (hunk 1 above)
Comment (jakule) [reply lib/srv/termmanager_test.go:31]: @xacrimon I don't think this test is correct as it works only on 32 bit systems. Look at this example: https://go.dev/play/p/FnTW_KRT__H on 64bit architecture it always passes even if you add something that breaks the alignment.

Comment (xacrimon) [reply lib/srv/termmanager_test.go:31]: Yep, missed that. Go will forcibly align it to 64-bit regardless on those platforms so this test ends up somewhat worthless.

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**